### PR TITLE
Fix types for sign-in options

### PR DESCRIPTION
**Pull request by shiatsumat**
Fix types for sign-in options.

### CODE REVIEW

File: externs/firebaseui-externs.js [hunk 1]
```diff
@@ -326,20 +330,30 @@ firebaseui.auth.SignInOption = function() {};
  */
 firebaseui.auth.SignInOption.prototype.provider;
 
+/**
+ * Defines the sign-in option
+ * needed to configure the FirebaseUI federated sign-in widget.
```
Comment (bojeil-google) [inline externs/firebaseui-externs.js]: Limit is 80 characters per line.
Try to fit as much as possible per line.

File: externs/firebaseui-externs.js [hunk 2]
```diff
@@ -231,7 +231,11 @@ firebaseui.auth.Config.prototype.signInFlow;
  * Determines the list of IdPs for handling federated sign-in, phone number as
  * well as password account sign-up.
  *
- * @type {!Array<!firebaseui.auth.SignInOption>|undefined}
+ * @type {!Array<
+ *  !firebaseui.auth.FederatedSignInOption|
+ *  !firebaseui.auth.EmailSignInOption|
+ *  !firebaseui.auth.PhoneSignInOption>
+ *  |undefined}
```
Comment (bojeil-google) [inline externs/firebaseui-externs.js]:
```
/**
 * ...
 *
 * @type {!Array<!firebaseui.auth.SignInOption|string>|undefined}
 */

File: externs/firebaseui-externs.js [hunk 3]
```diff
@@ -326,20 +330,30 @@ firebaseui.auth.SignInOption = function() {};
...
+/**
+ * Defines the sign-in option
+ * needed to configure the FirebaseUI federated sign-in widget.
+ *
+ * @interface
+ */
+firebaseui.auth.FederatedSignInOption = function() {};
+firebaseui.auth.FederatedSignInOption.prototype
+  = Object.create(firebaseui.auth.SignInOption);
+
```
Comment (bojeil-google) [inline externs/firebaseui-externs.js]: Replace with
```
/**
 * ...
 *
 * @interface
 * @extends {firebaseui.auth.SignInOption}
 */
firebaseui.auth.FederatedSignInOption = function() {};
```
You can use the same for `firebaseui.auth.EmailSignInOption` and `firebaseui.auth.PhoneSignInOption`

File: externs/firebaseui-externs.js (hunk 1 above)
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: Isn't this ok about line length limit?

**Comment by shiatsumat** [comment]
I'm not sure why travis-ci check failed...

File: externs/firebaseui-externs.js (hunk 1 above)
Comment (wti806) [reply externs/firebaseui-externs.js]: It's better to fit into the first line as much as possible:
 * Defines the sign-in option needed to configure the FirebaseUI federated
 * sign-in widget.

File: externs/firebaseui-externs.js [hunk 4]
```diff
@@ -231,7 +231,12 @@ firebaseui.auth.Config.prototype.signInFlow;
  * Determines the list of IdPs for handling federated sign-in, phone number as
  * well as password account sign-up.
  *
- * @type {!Array<!firebaseui.auth.SignInOption>|undefined}
+ * @type {!Array
```
Comment (wti806) [inline externs/firebaseui-externs.js]: Shall we use the signInOption interface here as @bojeil-google suggested?

File: externs/firebaseui-externs.js (hunk 1 above)
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: Oh, I got it.

File: externs/firebaseui-externs.js (hunk 4 above)
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: Typescript doesn't allow extra properties by default...
https://stackoverflow.com/questions/33836671/typescript-interface-that-allows-other-properties

File: externs/firebaseui-externs.js (hunk 3 above)
Comment (bojeil-google) [reply externs/firebaseui-externs.js]: Yeah seems fine.

File: externs/firebaseui-externs.js (hunk 1 above)
Comment (bojeil-google) [reply externs/firebaseui-externs.js]: It doesn't look like you changed it to @wti806's recommendation.

File: externs/firebaseui-externs.js (hunk 4 above)
Comment (bojeil-google) [reply externs/firebaseui-externs.js]: Closure externs should allow it. It is ok if the 2 (closure externs and typescript definitions) don't align. The end result will be the same.

File: externs/firebaseui-externs.js [hunk 5]
```diff
@@ -231,7 +231,8 @@ firebaseui.auth.Config.prototype.signInFlow;
  * Determines the list of IdPs for handling federated sign-in, phone number as
  * well as password account sign-up.
  *
- * @type {!Array<!firebaseui.auth.SignInOption>|undefined}
+ * @type {!Array<!firebaseui.auth.SignInOption|string>
```
Comment (wti806) [inline externs/firebaseui-externs.js]: Could you please put them on the same line?

Comment (shiatsumat) [reply externs/firebaseui-externs.js]: I did!

File: externs/firebaseui-externs.js [hunk 6]
```diff
@@ -392,14 +419,20 @@ firebaseui.auth.SignInOption.prototype.requireDisplayName;
  *   badge: (string|undefined)
  * }|undefined}
  */
-firebaseui.auth.SignInOption.prototype.recaptchaParameters;
+firebaseui.auth.PhoneSignInOption.prototype.recaptchaParameters = function() {};
```
[the remark this answers is not in the document]
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: I'm not sure what is the problem.
Is writing `firebaseui.auth.PhoneSignInOption.prototype.recaptchaParameters;` correct?

[the remark this answers is not in the document]
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: Got it. Maybe `PhoneSignInOption` instead of `SignInOption`...?

[the remark this answers is not in the document]
Comment (shiatsumat) [reply externs/firebaseui-externs.js]: OK, I did change!

**Pull request merged**